airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/extensions/LoggingInvocationInterceptor.java

```diff
@@ -110,25 +130,74 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
         LOGGER.error("Junit exception throw during {} after {}:\n{}", logLineSuffix, DurationFormatUtils.formatDurationWords(elapsedMs, true, true),
             stackTrace);
         throw t;
+      } finally {
+        timeoutTask.cancel();
+      }
+    }
+
+    private static class TimeoutInteruptor extends TimerTask {
+
+      private final Thread parentThread;
+      volatile boolean wasTriggered = false;
+
+      TimeoutInteruptor(Thread parentThread) {
+        this.parentThread = parentThread;
+      }
+
+      @Override
+      public void run() {
+        wasTriggered = true;
+        parentThread.interrupt();
       }
+
+      public boolean cancel() {
+        return super.cancel();
+      }
+
+    }
+
+    private static final Pattern PATTERN = Pattern.compile("([1-9]\\d*) ?((?:[nμm]?s)|m|h|d)?",
+        CASE_INSENSITIVE | UNICODE_CASE);
+    private static final Map<String, TimeUnit> UNITS_BY_ABBREVIATION;
+
+    static {
+      Map<String, TimeUnit> unitsByAbbreviation = new HashMap<>();
+      unitsByAbbreviation.put("ns", NANOSECONDS);
+      unitsByAbbreviation.put("μs", MICROSECONDS);
+      unitsByAbbreviation.put("ms", MILLISECONDS);
+      unitsByAbbreviation.put("s", SECONDS);
+      unitsByAbbreviation.put("m", MINUTES);
+      unitsByAbbreviation.put("h", HOURS);
+      unitsByAbbreviation.put("d", DAYS);
+      UNITS_BY_ABBREVIATION = Collections.unmodifiableMap(unitsByAbbreviation);
+    }
+
+    static Duration parseDuration(String text) throws DateTimeParseException {
+      Matcher matcher = PATTERN.matcher(text);
+      if (matcher.matches()) {
+        long value = Long.parseLong(matcher.group(1));
+        String unitAbbreviation = matcher.group(2);
+        TimeUnit unit = unitAbbreviation == null ? SECONDS
+            : UNITS_BY_ABBREVIATION.get(unitAbbreviation.toLowerCase(Locale.ENGLISH));
+        return Duration.ofSeconds(unit.toSeconds(value));
+      }
+      throw new DateTimeParseException("Timeout duration is not in the expected format (<number> [ns|μs|ms|s|m|h|d])",
+          text, 0);
     }
 
     private static Duration getTimeout(ReflectiveInvocationContext<Method> invocationContext) {
-      Duration timeout = DEFAULT_TIMEOUT;
+      Duration timeout = null;
       if (invocationContext.getExecutable()instanceof Method m) {
         Timeout timeoutAnnotation = m.getAnnotation(Timeout.class);
         if (timeoutAnnotation == null) {
           timeoutAnnotation = invocationContext.getTargetClass().getAnnotation(Timeout.class);
         }
         if (timeoutAnnotation != null) {
-          if (timeoutAnnotation.threadMode() == ThreadMode.SAME_THREAD) {
```

Comment:
This is a hack that was needed because some tests were using the ThreadLocal between @BeforeEach and @Test, and assertTimeoutPreemptively create a new thread for each call. So this new version uses the main Junit thread to execute test code and instead uses a different thread to be the interruptor